File: java/com/tigervnc/rfb/ZRLEDecoder.java
/* Copyright (C) 2002-2005 RealVNC Ltd.  All Rights Reserved.
 * Copyright (C) 2019 Brian P. Hinz
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
 * USA.
 */

package com.tigervnc.rfb;

import java.awt.image.*;
import java.nio.*;
import java.util.*;

import com.tigervnc.rdr.*;

public class ZRLEDecoder extends Decoder {

  private static ByteBuffer readOpaque24A(InStream is)
  {
    is.check(3);
    ByteBuffer r = ByteBuffer.allocate(4);
    r.put(0, (byte)is.readU8());
    r.put(1, (byte)is.readU8());
    r.put(2, (byte)is.readU8());
    return r;
  }

  private static ByteBuffer readOpaque24B(InStream is)
  {
    is.check(3);
    ByteBuffer r = ByteBuffer.allocate(4);
    r.put(1, (byte)is.readU8());
    r.put(2, (byte)is.readU8());
    r.put(3, (byte)is.readU8());
    return r;
  }

  public ZRLEDecoder() {
    super(DecoderFlags.DecoderOrdered);
    zis = new ZlibInStream();
  }

  public void readRect(Rect r, InStream is,
                      ServerParams server, OutStream os)
  {
    int len;

    len = is.readU32();
    os.writeU32(len);
    os.copyBytes(is, len);
  }

  public void decodeRect(Rect r, Object buffer,
                         int buflen, ServerParams server,
                         ModifiablePixelBuffer pb)
  {
    MemInStream is = new MemInStream((byte[])buffer, 0, buflen);
    PixelFormat pf = server.pf();
    switch (pf.bpp) {
    case 8:  zrleDecode8(r, is, zis, pf, pb); break;
    case 16: zrleDecode16(r, is, zis, pf, pb); break;
    case 32:
      {
        if (pf.depth <= 24) {
          int maxPixel = pf.pixelFromRGB(-1, -1, -1, pf.getColorModel());
          boolean fitsInLS3Bytes = maxPixel < (1<<24);
          boolean fitsInMS3Bytes = (maxPixel & 0xff) == 0;

          if ((fitsInLS3Bytes && pf.isLittleEndian()) ||
              (fitsInMS3Bytes && pf.isBigEndian()))
          {
            zrleDecode24A(r, is, zis, pf, pb);
            break;
          }

          if ((fitsInLS3Bytes && pf.isBigEndian()) ||
              (fitsInMS3Bytes && pf.isLittleEndian()))
          {
            zrleDecode24B(r, is, zis, pf, pb);
            break;
          }
        }

        zrleDecode32(r, is, zis, pf, pb);
        break;
      }
    }
  }

  private static enum PIXEL_T { U8, U16, U24A, U24B, U32 };

  private static ByteBuffer READ_PIXEL(InStream is, PIXEL_T type) {
    switch (type) {
    case U8:
      return ByteBuffer.allocate(1).put(0, (byte)is.readOpaque8());
    case U16:
      return ByteBuffer.allocate(2).putShort(0, (short)is.readOpaque16());
    case U24A:
      return readOpaque24A(is);
    case U24B:
      return readOpaque24B(is);
    default:
      return ByteBuffer.allocate(4).putInt(0, is.readOpaque32());
    }
  }

  private void zrleDecode8(Rect r, InStream is,
                           ZlibInStream zis,
                           PixelFormat pf, ModifiablePixelBuffer pb)
  {
    ZRLE_DECODE(r, is, zis, pf, pb, PIXEL_T.U8);
  }

  private void zrleDecode16(Rect r, InStream is,
                            ZlibInStream zis,
                            PixelFormat pf, ModifiablePixelBuffer pb)
  {
    ZRLE_DECODE(r, is, zis, pf, pb, PIXEL_T.U16);
  }

  private void zrleDecode24A(Rect r, InStream is,
                             ZlibInStream zis,
                             PixelFormat pf, ModifiablePixelBuffer pb)
  {
    ZRLE_DECODE(r, is, zis, pf, pb, PIXEL_T.U24A);
  }

  private void zrleDecode24B(Rect r, InStream is,
                             ZlibInStream zis,
                             PixelFormat pf, ModifiablePixelBuffer pb)
  {
    ZRLE_DECODE(r, is, zis, pf, pb, PIXEL_T.U24B);
  }

  private void zrleDecode32(Rect r, InStream is,
                            ZlibInStream zis,
                            PixelFormat pf, ModifiablePixelBuffer pb)
  {
    ZRLE_DECODE(r, is, zis, pf, pb, PIXEL_T.U32);
  }

  private void ZRLE_DECODE(Rect r, InStream is,
                           ZlibInStream zis,
                           PixelFormat pf, ModifiablePixelBuffer pb,
                           PIXEL_T pix_t)
  {
    int length = is.readU32();
    zis.setUnderlying(is, length);
    Rect t = new Rect();
    ByteBuffer buf = ByteBuffer.allocate(64 * 64 * pf.bpp/8);

    for (t.tl.y = r.tl.y; t.tl.y < r.br.y; t.tl.y += 64) {

      t.br.y = Math.min(r.br.y, t.tl.y + 64);

      for (t.tl.x = r.tl.x; t.tl.x < r.br.x; t.tl.x += 64) {

        t.br.x = Math.min(r.br.x, t.tl.x + 64);

        int mode = zis.readU8();
        boolean rle = (mode & 128) != 0;
        int palSize = mode & 127;
        ByteBuffer palette = ByteBuffer.allocate(128 * pf.bpp/8);

        for (int i = 0; i < palSize; i++) {
          palette.put(READ_PIXEL(zis, pix_t));
        }
        palette.flip();

        if (palSize == 1) {
          byte[] pix = new byte[pf.bpp/8];
          palette.get(pix);
          pb.fillRect(pf, t, pix);
          continue;
        }

        if (!rle) {
          if (palSize == 0) {

            // raw
            switch (pix_t) {
            case U24A:
            case U24B:
              ByteBuffer ptr = buf.duplicate();
              for (int i=0; i < t.area(); i++) {
                ptr.put(READ_PIXEL(zis, pix_t));
              }
              break;
            default:
              zis.readBytes(buf.duplicate(), t.area() * (pf.bpp/8));
            }

          } else {

            // packed pixels
            int bppp = ((palSize > 16) ? 8 :
                        ((palSize > 4) ? 4 : ((palSize > 2) ? 2 : 1)));

            ByteBuffer ptr = buf.duplicate();

            for (int i = 0; i < t.height(); i++) {
              int eol = ptr.position() + t.width()*pf.bpp/8;
              int b = 0;
              int nbits = 0;

              while (ptr.position() < eol) {
                if (nbits == 0) {
                  b = zis.readU8();
                  nbits = 8;
                }
                nbits -= bppp;
                int index = (b >> nbits) & ((1 << bppp) - 1) & 127;
                ptr.put(palette.array(), index*pf.bpp/8, pf.bpp/8);
              }
            }
          }

        } else {

          if (palSize == 0) {

            // plain RLE

            ByteBuffer ptr = buf.duplicate();
            int end = ptr.position() + t.area()*pf.bpp/8;
            while (ptr.position() < end) {
              ByteBuffer pix = READ_PIXEL(zis, pix_t);
              int len = 1;
              int b;
              do {
                b = zis.readU8();
                len += b;
              } while (b == 255);

              if (end - ptr.position() < len*pf.bpp/8) {
                System.err.println("ZRLE decode error\n");
                throw new Exception("ZRLE decode error");
              }

              while (len-- > 0) ptr.put(pix.array());

            }
          } else {

            // palette RLE

            ByteBuffer ptr = buf.duplicate();
            int end = ptr.position() + t.area()*pf.bpp/8;
            while (ptr.position() < end) {
              int index = zis.readU8();
              int len = 1;
              if ((index & 128) != 0) {
                int b;
                do {
                  b = zis.readU8();
                  len += b;
                } while (b == 255);

                if (end - ptr.position() < len*pf.bpp/8) {
                  System.err.println("ZRLE decode error\n");
                  throw new Exception("ZRLE decode error");
                }
              }

              index &= 127;

              ByteBuffer pix = ByteBuffer.allocate(pf.bpp/8);
              pix.put(palette.array(), index*pf.bpp/8, pf.bpp/8);

              while (len-- > 0) ptr.put(pix.array());
            }
          }
        }

        pb.imageRect(pf, t, buf.array());
      }
    }

    zis.flushUnderlying();
    zis.setUnderlying(null, 0);
  }

  private ZlibInStream zis;
}

File: common/rfb/ZRLEDecoder.cxx
/* Copyright (C) 2002-2005 RealVNC Ltd.  All Rights Reserved.
 * Copyright 2009-2022 Pierre Ossman for Cendio AB
 * 
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <rdr/InStream.h>
#include <rdr/MemInStream.h>
#include <rdr/OutStream.h>

#include <rfb/Exception.h>
#include <rfb/ServerParams.h>
#include <rfb/PixelBuffer.h>
#include <rfb/ZRLEDecoder.h>

using namespace rfb;

static inline uint32_t readOpaque24A(rdr::InStream* is)
{
  uint32_t r=0;
  ((uint8_t*)&r)[0] = is->readU8();
  ((uint8_t*)&r)[1] = is->readU8();
  ((uint8_t*)&r)[2] = is->readU8();
  return r;

}
static inline uint32_t readOpaque24B(rdr::InStream* is)
{
  uint32_t r=0;
  ((uint8_t*)&r)[1] = is->readU8();
  ((uint8_t*)&r)[2] = is->readU8();
  ((uint8_t*)&r)[3] = is->readU8();
  return r;
}

template<class T>
static inline T readPixel(rdr::ZlibInStream* zis)
{
  if (sizeof(T) == 1)
    return zis->readOpaque8();
  if (sizeof(T) == 2)
    return zis->readOpaque16();
  if (sizeof(T) == 4)
    return zis->readOpaque32();
}

static inline void zlibHasData(rdr::ZlibInStream* zis, size_t length)
{
  if (!zis->hasData(length))
    throw Exception("ZRLE decode error");
}

ZRLEDecoder::ZRLEDecoder() : Decoder(DecoderOrdered)
{
}

ZRLEDecoder::~ZRLEDecoder()
{
}

bool ZRLEDecoder::readRect(const Rect& /*r*/, rdr::InStream* is,
                           const ServerParams& /*server*/,
                           rdr::OutStream* os)
{
  uint32_t len;

  if (!is->hasData(4))
    return false;

  is->setRestorePoint();

  len = is->readU32();
  os->writeU32(len);

  if (!is->hasDataOrRestore(len))
    return false;

  is->clearRestorePoint();

  os->copyBytes(is, len);

  return true;
}

void ZRLEDecoder::decodeRect(const Rect& r, const uint8_t* buffer,
                             size_t buflen, const ServerParams& server,
                             ModifiablePixelBuffer* pb)
{
  rdr::MemInStream is(buffer, buflen);
  const rfb::PixelFormat& pf = server.pf();
  switch (pf.bpp) {
  case 8:  zrleDecode<uint8_t>(r, &is, pf, pb); break;
  case 16: zrleDecode<uint16_t>(r, &is, pf, pb); break;
  case 32: zrleDecode<uint32_t>(r, &is, pf, pb); break;
  }
}

template<class T>
void ZRLEDecoder::zrleDecode(const Rect& r, rdr::InStream* is,
                             const PixelFormat& pf,
                             ModifiablePixelBuffer* pb)
{
  int length = is->readU32();
  zis.setUnderlying(is, length);
  Rect t;
  T buf[64 * 64];

  Pixel maxPixel = pf.pixelFromRGB((uint16_t)-1, (uint16_t)-1, (uint16_t)-1);
  bool fitsInLS3Bytes = maxPixel < (1<<24);
  bool fitsInMS3Bytes = (maxPixel & 0xff) == 0;
  bool isLowCPixel = (sizeof(T) == 4) &&
                     ((fitsInLS3Bytes && pf.isLittleEndian()) ||
                      (fitsInMS3Bytes && pf.isBigEndian()));
  bool isHighCPixel = (sizeof(T) == 4) &&
                      ((fitsInLS3Bytes && pf.isBigEndian()) ||
                       (fitsInMS3Bytes && pf.isLittleEndian()));

  for (t.tl.y = r.tl.y; t.tl.y < r.br.y; t.tl.y += 64) {

    t.br.y = __rfbmin(r.br.y, t.tl.y + 64);

    for (t.tl.x = r.tl.x; t.tl.x < r.br.x; t.tl.x += 64) {

      t.br.x = __rfbmin(r.br.x, t.tl.x + 64);

      zlibHasData(&zis, 1);
      int mode = zis.readU8();
      bool rle = mode & 128;
      int palSize = mode & 127;
      T palette[128];

      if (isLowCPixel || isHighCPixel)
        zlibHasData(&zis, 3 * palSize);
      else
        zlibHasData(&zis, sizeof(T) * palSize);

      for (int i = 0; i < palSize; i++) {
        if (isLowCPixel)
          palette[i] = readOpaque24A(&zis);
        else if (isHighCPixel)
          palette[i] = readOpaque24B(&zis);
        else
          palette[i] = readPixel<T>(&zis);
      }

      if (palSize == 1) {
        T pix = palette[0];
        pb->fillRect(pf, t, &pix);
        continue;
      }

      if (!rle) {
        if (palSize == 0) {

          // raw

          if (isLowCPixel || isHighCPixel)
            zlibHasData(&zis, 3 * t.area());
          else
            zlibHasData(&zis, sizeof(T) * t.area());

          if (isLowCPixel || isHighCPixel) {
            for (T* ptr = buf; ptr < buf+t.area(); ptr++) {
              if (isLowCPixel)
                *ptr = readOpaque24A(&zis);
              else
                *ptr = readOpaque24B(&zis);
            }
          } else {
            zis.readBytes((uint8_t*)buf, t.area() * sizeof(T));
          }

        } else {

          // packed pixels
          int bppp = ((palSize > 16) ? 8 :
                      ((palSize > 4) ? 4 : ((palSize > 2) ? 2 : 1)));

          T* ptr = buf;

          for (int i = 0; i < t.height(); i++) {
            T* eol = ptr + t.width();
            uint8_t byte = 0;
            uint8_t nbits = 0;

            while (ptr < eol) {
              if (nbits == 0) {
                zlibHasData(&zis, 1);
                byte = zis.readU8();
                nbits = 8;
              }
              nbits -= bppp;
              uint8_t index = (byte >> nbits) & ((1 << bppp) - 1) & 127;
              *ptr++ = palette[index];
            }
          }
        }

      } else {

        if (palSize == 0) {

          // plain RLE

          T* ptr = buf;
          T* end = ptr + t.area();
          while (ptr < end) {
            T pix;
            if (isLowCPixel || isHighCPixel)
              zlibHasData(&zis, 3);
            else
              zlibHasData(&zis, sizeof(T));
            if (isLowCPixel)
              pix = readOpaque24A(&zis);
            else if (isHighCPixel)
              pix = readOpaque24B(&zis);
            else
              pix = readPixel<T>(&zis);
            int len = 1;
            int b;
            do {
              zlibHasData(&zis, 1);
              b = zis.readU8();
              len += b;
            } while (b == 255);

            if (end - ptr < len) {
              throw Exception ("ZRLE decode error");
            }

            while (len-- > 0) *ptr++ = pix;

          }
        } else {

          // palette RLE

          T* ptr = buf;
          T* end = ptr + t.area();
          while (ptr < end) {
            zlibHasData(&zis, 1);
            int index = zis.readU8();
            int len = 1;
            if (index & 128) {
              int b;
              do {
                zlibHasData(&zis, 1);
                b = zis.readU8();
                len += b;
              } while (b == 255);

              if (end - ptr < len) {
                throw Exception ("ZRLE decode error");
              }
            }

            index &= 127;

            T pix = palette[index];

            while (len-- > 0) *ptr++ = pix;
          }
        }
      }

      pb->imageRect(pf, t, buf);
    }
  }

  zis.flushUnderlying();
  zis.setUnderlying(nullptr, 0);
}
